base/src/main/java/org/mozilla/jss/netscape/security/x509/ExtensionsRequested.java

```diff
@@ -160,18 +160,12 @@ private void construct(DerValue dv) throws IOException {
         // now, the stream will be in the correct format
         stream.reset();
 
-        while (true) {
-            DerValue ext_dv = null;
-            try {
-                ext_dv = stream.getDerValue();
-            } catch (IOException ex) {
-                break;
-            }
-
-            Extension ext = new Extension(ext_dv);
+        DerValue extDv = null;
+        while (stream.available()>1) {
+            extDv = stream.getDerValue();
+            Extension ext = new Extension(extDv);
             exts.addElement(ext);
         }
```

Comment:
Hmm.. maybe this is the issue that I discussed with @ckelleyRH before. According to `InputStream.available()` [Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/InputStream.html#available()):
> Returns an **estimate** of the number of bytes that can be read (or skipped over) from this input stream **without blocking**, which may be 0, or 0 when end of stream is detected.

To my understanding this means that `available() <= 1` doesn't imply that it has reached the end of the stream. It's possible that there will be more data coming later. So terminating the loop based on this condition might not be correct.

Reply:
Yes, that is in the definition for the super class `InputStream`. However, in this case it is a ByteArrayInputStream which seems not have this behaviour, according to the [javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/ByteArrayInputStream.html#available()). In this case the `read()` is not blocking and the `available()` should return the length of all you can read, if I get correctly.

@edewata if you think my interpretation could be not correct let me know and I will follow the other approach with the `null` condition.

Reply:
Alternatively, we can move that logic into `getDerValue()` to hide the details from `DerInputStream` users. Something like this:
```
public DerValue getDerValue() throws IOException {

    // buffer.available() will return the remaining bytes in the stream precisely
    // (since it's a ByteArrayInputStream). DER values are at least 2 bytes long:
    // https://letsencrypt.org/docs/a-warm-welcome-to-asn1-and-der/

    return buffer.available() >= 2 ? new DerValue(buffer) : null;
}
```

Reply:
Comment added in the while loop.
The method `getDerValue()` is called from 91 different places and in all these cases the logic is based on the exception. If we modify the method to return `null` instead of the exception we need to change all the calling code and it is a big effort. I think this effort is valuable if we plan to reorganise these classes to optimise/modernise the code  but this will be in a separate PR.